### ISSUE TITLE
mosh prevents the use of scrollback

### ISSUE THREAD

**Issue by taralx** [mobile-shell/mosh#122]
If I use mosh to connect, and then run a command with a lot of output, e.g. ps, my scrollback buffer doesn't get any of the overflow. This is sad-making -- I don't always know when a command will produce a lot of output, and, additionally, I like scrollback for seeing previous commands' output.

**Comment by MichaelSp** [comment]
Here is my +1 if this makes any difference....

**Comment by The-Compiler** [comment]
Someone contributing code would probably make a difference :wink:

But to be honest, I don't see the point of implementing this in mosh - if you want native client-side scrolling you might (as it's been said) basically just use ssh/autossh/screen or tmux to get  (nearly) the same behavior.

If you want server-side scrolling, you can already have it today by using screen/tmux on the server, so I don't think it makes sense to duplicate that feature in mosh.

**Comment by hgbork** [comment]
Within xfce4-terminal a 
   mosh user@example.com --no-init 
keeps scrolling as it is and I see everything, even the output I had before invoking mosh. 

But yes, this works only if the output is generated line by line and in a huge chunk; e.g. (quite useless command)
  while read l; do echo "$l"; sleep 0.01; done < <(ls -alsR /etc)

For me this works fine and if commands produce big chunks of text in short time (and mosh cut them), I can either pipe to $PAGER or send it to a file.

**Comment by ThomasSteinbach** [comment]
Mhh :/ Hope it will ever be implemented

+1

**Comment by scottj97** [comment]
My understanding is that mosh starts dropping output when it sees that the client end can't keep up. That results in an incomplete native scroll buffer on the client end.

Could it be as simple as adding a flag to disable this output dropping feature? --no-drop?

**Comment by andersk** [comment]
@scottj97 No. There’s no block of code somewhere that implements an “output dropping feature” to be disabled. Output dropping is an emergent property of the [Mosh design](https://mosh.mit.edu/mosh-paper.pdf) and protocol. Enabling reliable scrollback will require substantial changes to both.

**Comment by gnowxilef** [comment]
very much +1

**Comment by briner** [comment]
I'd love to have this feature. I'll be ready to pay for that. Why not create a bounty for it ?

**Comment by jordanade** [comment]
It sounds like most would be happy without "complete native scrollback" but rather, say, 3x or some configurable ratio of the current screen's height. But it sounds like mosh couldn't do this without communicating the false 3x height to apps like vi and emacs, thus ruining their full-screen UI. There's no way to tell the server that the screen is 24 rows while actually syncing 72 rows. Is that right?

**Comment by cgull** [comment]
Correct. The emulated terminal's window size must be the same as that displayed for full screen apps to work well. The way scroll back is usually implemented is that an emulator maintains a separate buffer of rows for scroll back. When cursor motion pushes text off the window it's copied into the scroll back buffer.  So the size doesn't matter much for implementation.

But on Mosh, we have to do that on the server side and then find some way to display it on client, either in band from the remote server (but screen/tmux do this well already) or out of band, on the local client, with an extension to the protocol (big task, and it still wouldn't get native-like scroll back in your outer GUI terminal emulator).  It's not an easy/trivial problem to fix

**Comment by gvaish** [comment]
Though not really a perfect solution, but using `screen` helps me scrollback.

Using the default bindings for `screen`:

. Execute `C-a Esc` to enter into [copy / block selection](https://www.gnu.org/software/screen/manual/screen.html#Copy) mode.
. Use mouse scroll, up/down or `C-b`/`C-f` to scroll.

**Comment by MisterTea** [comment]
If you are willing to sacrifice type prediction, you can use Eternal Terminal.  ET keeps your connection alive and supports native scrollback: https://mistertea.github.io/EternalTCP/  It's good if you have a low-latency connection and are using mosh for the persistence.

**Comment by ezyang** [comment]
https://github.com/4ast/mosh/commit/b642df68efe1b437ceeb6ccb2d7369b4f9518fd6 may be of interest

**Comment by rerime** [comment]
Mrrr.....Hope it will ever be implemented
+100500

**Comment by sterpe** [comment]
Guessing this is still not resolved...

**Comment by roosemberth** [comment]
+1, I think people would even like a bounty for this

**Comment by danielmotaleite** [comment]
how about having a mosh option to automatically start screen or tmux... this would allow both the mosh features and the screen/tmux scrollback!

I think most people really miss the scrollback, but always forget to start screen or tmux ... if mosh could do that automatically, the problem is solved

**Comment by t-nelson** [comment]
@danielmotaleite mosh doesn't need to support that.  You can do it yourself.  I've achieved the same for years with the following for in ~/.profile
```
if [[ "$SSH_TTY" != "" ]]
then
  if [[ "$TERM" != "screen" ]]; then
    screen -dr default
    if (($?)); then
      screen -ln -S default
    fi
  fi
fi
```

**Comment by calder** [comment]
+large-number. This is a dealbreaker.

**Comment by cpakkala** [comment]
I appreciate all the hard work on mosh, but the "workaround" is simply not good enough.  No one has mentioned it, but I find it hard to believe that I'm the only one whose use-case is as follows:  

Start a **single** tmux session on one host, per activity ("session" and "activity" to me are synonymous), and then use mosh within various tmux panes to remote into other machines.  This is much cleaner and more organized than moshing into each machine and then starting a tmux session on each one.   More importantly doing the latter breaks the abstraction of an "activity" for me; yet the former is problematic because then I either have to deal with the missing scrollback in each of the mosh panes, or I have to nest a tmux session in each pane.   Nesting, of course causes many larger issues and is to be avoided...

**Comment by ddacunha** [comment]
I believe I have read most of the 67 previous comment I appreciate this issue is still open so that mosh can gracefully manage packet loss handling and have an always working Control-C.
However I have not see the following proposal: could I choose how many lines mosh actually returns to my terminal upon completion of the command? Or is that not possible without the terminal app allowing to change the terminal history?

**Comment by rledisez** [comment]
I rebased patch from @4ast on v1.3.2 (see https://github.com/rledisez/mosh/commits/localScrollback-1.3.2)

@andersk if I take time to make it an option that must be enabled on the client side, would this be something that could be accepted eventually?

At this point, I'm just asking if it is worth I spend time on it (if yes, i take any advices ;))

**Comment by NoahO** [comment]
@rledisez even if you have to fork mosh, you will still have users. I can't use eternalTCP because I need UDP support.

**Comment by netroby** [comment]
+1 scrollback

**Comment by crushcafeteria** [comment]
please give us scrollback. pleaseeeeee.....

**Comment by Sora233** [comment]
+1

**Comment by Jan69** [comment]
I can't believe this has dragged on for 6 years, at this point, and still nowhere near the answer....
also, unrelated to this specific issue, but there's no way to even just disable all of the history-stuff? why not?

**Comment by tomas** [comment]
@rledisez does scrollback work in your branch? if there's anything missing just point me in the right direction and I'll try to get it working.

**Comment by shovas** [comment]
+1 Scrollback needed. Screen or tmux are effective, I imagine, but I don't think I'd make the switch without this being native. I just don't like have to run extra programs every time I login. I have too many sessions for that.

**Comment by KopfKrieg** [comment]
> I have too many sessions for that.

How about one tmux session per machine? You can create new "windows" within tmux, or split the view.

**Comment by tomas** [comment]
I'm trying out @rledisez's branch and, although type prediction isn't available, native scrollback works beautifully.

**Comment by abhatia2107** [comment]
@rledisez As it looks like native scrollback is working in your branch. Can we expect it to come in master branch?

**Comment by NyaMisty** [comment]
Why there aren't any feedback from the maintainers?
This issue has been opened for almost 6.5 years, and someone has made a patch. 
Whether merge it or not, plz give us some response ;)
I'm hoping this will be resolve before its 7th anniversary.

**Comment by LvChengbin** [comment]
+ The weight of the world.

**Comment by AndrewSav** [comment]
>I'm hoping this will be resolve before its 7th anniversary.

Nope. And by the looks of it one should not hold their breath even for 10th.

**Comment by notmike-5** [comment]
sad

**Comment by dandv** [comment]
For those who want autossh + mosh + native scrolling, there's Eternal Terminal.

![image](https://user-images.githubusercontent.com/33569/61596071-9c393000-abb3-11e9-85e6-700b323330b6.png)

`mosh` is a great idea, but I need the scrollback buffer a lot more than I need local echo, so I'm very happy with ET.

**Comment by sksd509** [comment]
Is this still not supported?
I have recently started using this combination: mosh + iterm + tmux.
Can I attain a native scrolling feature by doing some magic within their configurations?

**Comment by eminence** [comment]
Correct, mosh does not have any ability to manage a local scrollback buffer.  Your best option remains to use tmux or screen on the remote system.  Some options (like setting `mouse on` in tmux) will let you  use your mouse wheel to automatically scroll the tmux buffer.

**Comment by zhouzach** [comment]
@eminence I use tmux on mac, but can not attain a native scrolling feature.why?

**Comment by dandv** [comment]
> Your best option remains to use tmux or screen on the remote system.

Another option is to use Eternal Terminal. It manages the local scrollback buffer very well.

/unsubscribe

**Comment by notmike-5** [comment]
Soylent green is people

**Comment by axeII** [comment]
Eight years later.... 

Umm any update on this 🙏? This is really one of _the_ feature I am missing on mosh 😞.

**Comment by RipperSK** [comment]
I was looking into using Mosh. I have it installed and tested for the most part, but this is a show stopper for me :(

Is there a way to get this feature implemented ?

**Comment by sksd509** [comment]
> Is this still not supported?
> I have recently started using this combination: mosh + iterm + tmux.
> Can I attain a native scrolling feature by doing some magic within their configurations?

I am able to scroll somehow by using tmux with few extra configs in `~/.tmux.conf`. It's already mentioned above by other users. Recommend others to follow the suit if possible.

**Comment by marty1885** [comment]
@RipperSK The root cause of not able to scroll is caused by `mosh` entering a "alternative screen" mode. It can be disabled using the `--no-init` flag ([see the doc](https://github.com/mobile-shell/mosh/blob/60859e9d090f530022ffc3e1345975bfdd1b3978/man/mosh.1#L227-L230)). Just `alias mosh='mosh --no-init'` and the problem will go away.

**Comment by Kabouik** [comment]
> @RipperSK The root cause of not able to scroll is caused by `mosh` entering a "alternative screen" mode. It can be disabled using the `--no-init` flag ([see the doc](https://github.com/mobile-shell/mosh/blob/60859e9d090f530022ffc3e1345975bfdd1b3978/man/mosh.1#L227-L230)). Just `alias mosh='mosh --no-init'` and the problem will go away.

Beautiful. Could someone please explain in Layman's terms what are the drawbacks?

**Comment by eminence** [comment]
> The root cause of not able to scroll is caused by mosh entering a "alternative screen" mode.

This is not true.  The root cause is that the mosh model is fundamentally incompatible with a scroll-back buffer managed by your local terminal.

> Beautiful. Could someone please explain in Layman's terms what are the drawbacks?

The drawback is that your local scrollback might be "incomplete".  For example, try `cat`ing a large text file.  With mosh, probably only the last page of the file will appear in your local scrollback.  If you use ssh, you'll have the entire file in your local scrollback.  This is a feature of mosh (not a bug)

**Comment by AndrewSav** [comment]
> The drawback is that your local scrollback might be "incomplete". For example, try cating a large text file. With mosh, probably only the last page of the file will appear in your local scrollback.

What does this workaround achieve then?

**Comment by andersk** [comment]
@AndrewSav Your question has been [addressed in earlier comments](https://github.com/mobile-shell/mosh/issues/122#issuecomment-32106757). (I don’t mean to be rude—it’s just the more we repeat the previous discussion, the harder it gets to read the previous discussion.)

**Comment by AndrewSav** [comment]
@andersk I read that now but it does not seem to be related to the workarond. Also it seems to be way up before the workaround was posted, which also hints it could be unrelated. What am I missing?

**Comment by Khalian** [comment]
A much simpler alternative to mosh would be https://www.harding.motd.ca/autossh/

Combine that with tmux/screen/byobu and its a lot less hassle to maintain. No extra servers to install on the service machine either, just plain old ssh.

**Comment by avatar-lavventura** [comment]
> @RipperSK The root cause of not able to scroll is caused by `mosh` entering a "alternative screen" mode. It can be disabled using the `--no-init` flag ([see the doc](https://github.com/mobile-shell/mosh/blob/60859e9d090f530022ffc3e1345975bfdd1b3978/man/mosh.1#L227-L230)). Just `alias mosh='mosh --no-init'` and the problem will go away.

@marty1885 I have added 'mosh --no-init'` but it prevents use of mouse wheel in the `emacs`, it is possible to enable the mouse wheel while using `--no-init`?

**Comment by marty1885** [comment]
@eminence Seeing all the comment and feature requests. I think mosh might want to look into implementing it. IMO most people just want a ssh alternative that automatically reconnects after a outage. A broken/terminal with some extra text is still far better than a dead ssh session. I'm not savvy enough to write it. If you think it's worth a try.. May you point me in a direction? I'm sure there's a way since `ssh` does it.

@avatar-lavventura I did some test. And it seems that `smcpu` and `rmcup` sequences used by emacs doesn't work under mosh with `--no-init`. But vim scrolling still works. You might want to submit a feature request to emacs?

**Comment by avatar-lavventura** [comment]
Hm on my end it does not work in `nano`, `less`, `vim`, `vi` :-( I mean it scrolls back in the terminal history but it does not scroll down inside the `less` or any other editor like `emacs`. 

For example, in `ssh`, when I do `less file`, scroll up will go to highest point in the opened file, but with `mosh --no-init` it goes to top point of the terminal history.

=> Does it work in `less` or `nano` on your end?

**Comment by eminence** [comment]
Hi all, I am marking the above comments as "off topic".   I want to reiterate a few things:

* I don't consider using `--no-init` as a way to get reliable access to a local scrollback buffer.  You can of course use this option if you understand that it does and its limitations, but if you're having problems with `--no-init`, this issue is not the place to discuss those issues.  Please open a new issue, or find us on IRC to chat
* We definitely know that people want this feature, it's one of the most requested features.   But just because ssh is designed in a way to support a local scrollback doesn't mean that mosh can make use of this.    I know we've left this issue open for many years, and this might have signaled that we intend to implement this feature.   But please interpret this open issue as simply an acknowledgement of the request.   **We do not have any intention on implementing this feature**
* This issue likely has hundreds of subscribers.   Please think *very carefully* before commenting -- does the comment/question you are about to add substantially contribute to the conversation?  If not, please open a [new issue](https://github.com/mobile-shell/mosh/issues/new) with your question.

Thanks all

**Comment by avatar-lavventura** [comment]
@eminence The issue was open since 2012, which is around 8 years and I believe it is apponded. If it is implemented it was implemented by now. There is not hundreds of subscribers in this thread there is only 77. Does your comment contributed for a solution? Using `tmux or screen` also does not provide a solution for this. 

I am not having issue with `--no-init`, I am having issue with `mosh` , which does not support scroll back in terminal history along within tools like `emacs, nano, less` and so on. If a developer swithcing from `ssh` to `mosh` there is no way he/she can adapt the scroll behavior of `mosh`. Without having both together like `ssh` does, there is no meaning to use `mosh` at all.

=> https://github.com/mobile-shell/mosh/issues/1110
=> https://stackoverflow.com/questions/63616612/in-mosh-how-to-scrollback-in-the-terminal-and-use-mouse-in-emacs-less-etc

**Comment by avatar-lavventura** [comment]
Since comments are off topic please see my question at StackOverflow related to `how can I scrollback in the terminal and use mouse in emacs, less etc`: https://stackoverflow.com/q/63616612/2402577

**Comment by FFAMax** [comment]
Use case: After working from location with a high speed Internet took the laptop and moved to bus.
What changed? Right, connectivity: from stable and high speed to -> unstable and sometimes low speed.
Therefore, if we talk not about ignoring all these comments but about hearing and follow to the initial logic of mosh, I see we need another mode. Mode which will determine behaviours. If we can pass all data to client (high speed and stable connection) - why not do that? If mosh detected connection got jitter/delays/losses then -> switch to legacy mode. If connection restored -> switch to high performance mode. Also may be useful with additional options like --mode-hybrid if users really want use this feature. Because I see variants when people may wanna save unneeded traffic even with a stable connection (for example in airplane, roaming, etc).

btw: tmux is another - this require exchange with a remote side.
Example: I worked from workplace with a stable connection, had same scroll buffer, after jumped to the bus - if scroll buffer not on my local terminal, it will relay on network reachability instead of to be available locally. :)

**Comment by achernya** [comment]
Unfortunately the problem here is not network communication, but rather mosh's construction of being a terminal emulator. Native scrollback requires not just a reliable channel, but also not being a full terminal emulator.

This issue is marked as a feature request and remains open in the mosh bug tracker slated for some future milestone. This means that the mosh team has an interest in working on this, but is not committing to a timeline for its completion.

To folks that are interested in this feature: we're interested in proposals to fix this and PRs to do so, but please don't comment on this issue attempting to motivate the use case further. Rest assured that you've been heard -- feel free to 👍 up the top comment to express your support.

**Comment by eternaleye** [comment]
> Native scrollback requires not just a reliable channel, ***but also not being a full terminal emulator.***
(emphasis mine)

I'm not sure this is strictly true? Specifically, as far as I'm aware it's more a matter of what specific terminal one is emulating. For example, the vt520 has a "desktop feature" to "Review previous lines", this keeps a "Review Previous Lines Buffer", which (quoting section 2.8.3 of the EK-VT520-RM document) behaves as follows:

> The Review previous lines feature when selected re-arranges page memory into a single page and a scroll back buffer (transcript). Lines of text scrolled up from the first line of the page are saved in the scroll back buffer (replacing the oldest lines saved if any). Pressing Control with the up or down arrow keys pans the screen window up or down allowing you to view lines previously scrolled off the top of the page up to the number of lines saved in the scroll back buffer. If the page size is larger than the window size, pressing Control with the up arrow key first pans to the top of the page and then pans into the scroll back buffer.
> 
> Any time lines in the scroll back buffer are available, a small down and/or up arrow appears at the lower right position of the status line to indicate there are additional unseen lines of the page below or above. When any text is received from the host for this session, the screen window is automatically repositioned to show the single page before the display is updated. Switching sessions does not automatically reposition the screen allowing you to view the scroll back data in one window while working in another. The number of lines that can be saved in the scroll back buffer is one page less than the total number of lines (in 25 line pages) allocated to the session virtual terminal. The factory default for Review previous lines is disabled.
> 
> You cannot copy and paste from the scroll back buffer.

As far as I can tell, this functionality provides perfect license for a VT520 emulator to provide a line-by-line scrollback buffer. I've not yet gone digging enough to tell when such functionality was first introduced (i.e. whether it predates the VT520).

As for the first half (reliable channel), I'm not convinced of that either: in particular, the _visible portion_ of the scrollback buffer could be the "synchronized state", while the scrollback buffer itself could be "server-private state". At that point, you no longer need a reliable channel either.

**Comment by achernya** [comment]
The problem is not the VT520 that mosh emulates. The problem is the outer terminal requires a line-oriented output, of every single line. If you want "native scrollback", then mosh needs to deliver, reliably, every single event that happened server-side to the client's outer terminal.

The scrollback buffer can't be synchronized in the way you describe because the scrollback buffer is part of the outer terminal. There are no events (to my knowledge) that the outer terminal sends to mosh to generate scrollback; that's why things like screen and tmux have their own scrollback mechanisms. What you're asking for essentially amounts to making a "moshterm" GUI program.

**Comment by zen2** [comment]
As far as I have understood mosh implementation uses two buffer: one on client and one on server side and the sync mechanism is using a diff/patch of both buffers to keep in sync.
So I suppose that the buffer size is equal to the number of characters displayed on your current terminal so basically height x width characters.
On another side, I can see that mosh doesn't interpret the classical keys input of shift+PageUp/PageDown natively so these keys are not used at all since it is not even treated natively by readline library.

So I wonder if it is not possible to implement a scrollback by pages simply by setting the buffer size by multiplying the size of the buffer by an integer in order to make the buffer represent a finite number of full pages for the current terminal and interpret keys to change page by page or line by line. Like it the sync mechanism continue to do is job and a simple offset applied permit to display the scrollback. The drawback is that won't work at all if you resize the terminal but actually mosh doesn't handle it too for its buffer so I don't think it is a real problem any way.

It is a similar that the way used for old school vertical bitmap scrolling where you simple change offset of the video memory pointer on a virtual screen more higher than real screen.

**Comment by AkechiShiro** [comment]
No solution was found possible for this I guess ? as I've read @achernya's latest comment right ? Maybe this feature is simply not possible and we should use screen/or any other terminal multiplexer for scrollback buffer ?

**Comment by otomn** [comment]
Here is another solution:
```bash
mosh
vim +"ter ++curwin"
```
then you can do `ctrl-w``N` to enter normal mode where you can scroll freely, and `i` to return to typing commands.
Since this is vim, you can customize the key binds easily.
